### PR TITLE
Generalize "CreateFMix"

### PR DESCRIPTION
**Pull request by jiaxie-amd**
Create "fmix" operation, returning ( 1 - A ) * X + A * Y. Result would
be FP scalar or vector value.

    - Returns scalar, if and only if "X", "Y" and "A" are all scalars.
    - Returns vector, if "X" and "Y" are vector but "A" is a scalar,
      under such condition, "A" will be splatted.
    - Returns vector, if "X", "Y" and "A" are all vectors.

Note that when doing vector calculation, it means add/sub are element-wise
between vectors, and the product will be Hadamard product.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 235c56b5ea50cb3d69cafc1d5a7483e575193726
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 9fab16015e522fff05890a045a1e9d8d3c23a636</li>
<li>SPVGEN: dev a61db14f75728e91c243d4bfc1c39700e9dc0ba6</li>
<li>XGL: dev 052ebfa2e77abd14fdb4f5120b35b081f3fc4bdc</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ecf0a1496f32eaf34825b3d11432c6b1d03c951</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 942/942 (100.0%)</li>
<li>Failed: 0/942 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 65463/65463 (100.0%)</li>
<li>Failed: 0/65463 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit f23a5db374eddf7505c3566faab18ea00cba8696
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 9fab16015e522fff05890a045a1e9d8d3c23a636</li>
<li>SPVGEN: dev a61db14f75728e91c243d4bfc1c39700e9dc0ba6</li>
<li>XGL: dev 052ebfa2e77abd14fdb4f5120b35b081f3fc4bdc</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ecf0a1496f32eaf34825b3d11432c6b1d03c951</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 942/942 (100.0%)</li>
<li>Failed: 0/942 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 65463/65463 (100.0%)</li>
<li>Failed: 0/65463 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/builder/llpcBuilderImplArith.cpp [hunk 1]
```diff
@@ -1329,6 +1329,40 @@ Value* BuilderImplArith::CreateFindSMsb(
...
+
+    Value* pYSubX = CreateFSub(pY, pX);
+    if (auto pVectorResultTy = dyn_cast<VectorType>(pYSubX->getType()))
+    {
+        // pX, pY => vector, but pA => scalar
+        if (pA->getType()->getScalarType()->isFloatTy())
+        {
+            pA = CreateVectorSplat(pVectorResultTy->getVectorNumElements(), pA);
+        }
+    }
```
Comment (amdrexu) [inline llpc/builder/llpcBuilderImplArith.cpp]: Add a blank line afte this line

File: llpc/builder/llpcBuilderImplArith.cpp [hunk 2]
```diff
@@ -1329,6 +1329,40 @@ Value* BuilderImplArith::CreateFindSMsb(
...
+    FastMathFlags ModifiedFMF = getFastMathFlags();
```
Comment (amdrexu) [inline llpc/builder/llpcBuilderImplArith.cpp]: ModifiedFMF -> fastMathFlags

File: llpc/builder/llpcBuilderImplArith.cpp [hunk 3]
```diff
@@ -1329,6 +1329,40 @@ Value* BuilderImplArith::CreateFindSMsb(
...
+// Returns vector, if "pX", "pY" and "pA" are all vectors.
+// Note that when doing vector calculation, it means add/sub are element-wise between vectors, and the product will
+// be Hadamard product.
+Value* BuilderImplArith::CreateFMix(
+    Value*        pX,        // [in] left Value
+    Value*        pY,        // [in] right Value
+    Value*        pA,        // [in] wight Value
+    const Twine& instName)   // [in] Name to give instruction(s)
+{
+    Value* pResult = nullptr;
```
Comment (amdrexu) [inline llpc/builder/llpcBuilderImplArith.cpp]: Remove this and replace pMixed with pResult.

File: llpc/builder/llpcBuilderImplArith.cpp [hunk 4]
```diff
@@ -1329,6 +1329,40 @@ Value* BuilderImplArith::CreateFindSMsb(
...
+        // pX, pY => vector, but pA => scalar
+        if (pA->getType()->getScalarType()->isFloatTy())
+        {
+            pA = CreateVectorSplat(pVectorResultTy->getVectorNumElements(), pA);
+        }
+    }
+    FastMathFlags ModifiedFMF = getFastMathFlags();
+    ModifiedFMF.setNoNaNs();
+    ModifiedFMF.setAllowContract();
+    CallInst* pMixed = CreateIntrinsic(Intrinsic::fmuladd, pX->getType(), {pYSubX, pA, pX}, nullptr, instName);
```
Comment (amdrexu) [inline llpc/builder/llpcBuilderImplArith.cpp]: pMixed -> pResult

**Comment by jiaxie-amd** [comment]
> Check SPIRVToLLVM::transGLSLExtInst(), and replace original implementation by using builer's CreateFMix(). Maybe some shaderdb test files need modification (/llpc/llpc/test/shaderdb).
> 
> ```
> case GLSLstd450FMix: {
>     // Linear blend
>     Value *X = Args[0];
>     Value *Y = Args[1];
>     Value *A = Args[2];
>     Value *Diff = getBuilder()->CreateFSub(Y, X);
>     return getBuilder()->CreateIntrinsic(Intrinsic::fmuladd, X->getType(),
>                                          {Diff, A, X});
>   }
> ```

Oh, I just forget to add this file's change in the commit.  I'll test if it need modification in shaderdb tests.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9658cc2d279cc171fba65041492218b69af9ec0d
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 9fab16015e522fff05890a045a1e9d8d3c23a636</li>
<li>SPVGEN: dev a61db14f75728e91c243d4bfc1c39700e9dc0ba6</li>
<li>XGL: dev 052ebfa2e77abd14fdb4f5120b35b081f3fc4bdc</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ecf0a1496f32eaf34825b3d11432c6b1d03c951</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 942/942 (100.0%)</li>
<li>Failed: 0/942 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 65463/65463 (100.0%)</li>
<li>Failed: 0/65463 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit ce2ed2e16b5b15bbf741bc5fd468cb1278af1278
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 9fab16015e522fff05890a045a1e9d8d3c23a636</li>
<li>SPVGEN: dev a61db14f75728e91c243d4bfc1c39700e9dc0ba6</li>
<li>XGL: dev 052ebfa2e77abd14fdb4f5120b35b081f3fc4bdc</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ecf0a1496f32eaf34825b3d11432c6b1d03c951</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 942/942 (100.0%)</li>
<li>Failed: 0/942 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 65463/65463 (100.0%)</li>
<li>Failed: 0/65463 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**